### ISSUE TITLE
Support hooks (Before, BeforeAll, After, AfterAll)

### ISSUE THREAD

**Issue by mortelee** [vitalets/playwright-bdd#15]
As already described hooks are currently not supported.
When generating the  spec files from the feature files you can also not know that hooks are defined.

Is it an option to add a config option where you can define the location of the file(s) with hooks?
In that way you can add hooks to the generated spec files.

**Comment by vitalets** [comment]
Interesting idea. So mean convert Cucumber hooks into Playwright hooks?

**Comment by mortelee** [comment]
![image](https://user-images.githubusercontent.com/1556873/233963900-fe09e463-9bfd-4bc0-83d7-5d55d5be7579.png)

![image](https://user-images.githubusercontent.com/1556873/233964945-fadc769f-9f7e-4a05-a9cd-76e7e73bad8d.png)

Like the Given, When, Then steps you like to be able to call the BeforeAll, Before, AfterAll, After cucumber hooks from the test.beforeAll, test.beforeEach, test.afterEach,  test.AfterAll functions.

Would it be possible to pass the scenario object from cucumber on the before and after hooks?

Not sure if it will be possible to call the multiple After hooks since in cucumber you can define multiple hooks of the same type and they will be executed in the order they are defined, but I don't have a problem to put everything in one After hook for now.

Becomes different when you also support tags and you can make it such that a hook is only called when the scenario has a certain tag. But thats for later i would say.

**Comment by mortelee** [comment]
Never tried it with fixtures before, but playwright documentation explains some advantages. Maybe worthwhile to investigate.

**Comment by paolodg** [comment]
this would be a nice feature for me too

**Comment by rcruz11-rbi** [comment]
This feature would be great

**Comment by gzuppl** [comment]
I stumbled upon a related problem, trying to port a suite of tests (using wdio and cucumber) to playwright.

There are a handful of test cases in that suite with potentially interruptive side effects. I used [tags + conditional hooks](https://cucumber.io/docs/cucumber/api/?lang=javascript#conditional-hooks) to assure those state changes are reverted when those tests fail/pass.

Maybe im missing something but as far as i know there is no way i can define an After-Hook for a specific test-case in playwright-bdd currently. The playwright-way of doing this would be done in generated test-case-files and the cucumber-way relies on hooks + tags.

**Comment by vitalets** [comment]
> Maybe im missing something but as far as i know there is no way i can define an After-Hook for a specific test-case in playwright-bdd currently. The playwright-way of doing this would be done in generated test-case-files and the cucumber-way relies on hooks + tags.

I think the most playwright-way for that task is to use [fixtures](https://playwright.dev/docs/test-fixtures). 
Let me show an example.

Imagine a scenario with a step that requires authorization:
```feature
Feature: Some feature

    Scenario: Some scenario
        Given I am an authorized user
```
So we need to sign-in user before running step and sign-out after the step.


**In Cucumber** I can add some tag (e.g. `@auth`) to that scenario and add Before and After conditional hooks:
```feature
Feature: Some feature

    @auth
    Scenario: Some scenario
        Given I am an authorized user
```
Hooks:

```ts
Before({tags: '@auth'}, async function () {
  console.log('sign-in');
});

After({tags: '@auth'}, async function () {
  console.log('sign-out');
});
```

**In Playwright** I can solve it with the following `auth` fixture:
```ts
export const test = base.extend({
  auth: async ({}, use) => {
    console.log('sign-in');
    await use({ username: 'some user' });
    console.log('sign-out');
  }
});
```
And just reference `auth` fixture in the step definition:
```ts
Given('I am an authorized user', async ({ auth }) => {
  console.log('step for authorized user', auth.username);
});
```

The benefits of using fixture:
- No need for extra tag
- If authorized step is removed from scenario, sing-in and sing-out are automatically skipped
- `auth` fixture is reusable in other steps 

I understand that you may have existing tests implemented with hooks, so I'm still thinking how to supporting them.
But also you can share your case here and we will try to implement it with fixtures.

**Comment by gzuppl** [comment]
Thanks a lot for your explaination! Sadly it looks like I 'can not' use the approach you pointed out in combination with what i have been doing with a [custom world](https://github.com/cucumber/cucumber-js/blob/main/docs/support_files/world.md#custom-worlds):

To deal with unique fields and potential data-collisions between tests i make extensive use of storing uuids in that custom world to access in following steps. That way i can leave out those details in my tests, not worry about collisions and reference the thing i created in later steps. The following two steps from my codebase-in-transition might help illustrate what i mean:

`Given<SamplesWorld>(/^a SampleType$/, async function () 
{
    const label = crypto.randomUUID();
    const handle = crypto.randomUUID();
    
    await createSampleTypeIfNotExisting({ label, handle });
    this.sampleType = { label, handle };
});`

`When<SamplesWorld>(/^I edit the given SampleType setting the (?:Permitted Unit|Permitted Units):$/,
    async function (newUnits: DataTable)
{
    const permittedUnits = newUnits.transpose().raw()[0];
    
    await openViewAndEditSampleType(this.sampleType.label);
    await updateSampleTypeProps({ permittedUnits });

    await editView.returnToLastVisitedView();
});`

Because of that i feel like i am locked into using your [cucumber-steps-style](https://vitalets.github.io/playwright-bdd/#/writing-steps?id=cucumber-style) which your proposed solution does not seem to be compatible with: 
![image](https://github.com/vitalets/playwright-bdd/assets/67320193/c0aa9367-4a74-40ce-916c-ee64582e4cdb)

Lets say.. id much rather loose hooks ;).

**Comment by vitalets** [comment]
@gzuppl you are right, using custom fixtures with cucumber-style is a gap now. I should think about it.. 
Btw, could you share the code of `SamplesWorld` (if possible) ?

**Comment by gzuppl** [comment]
Basically it just exposed a bunch of setters and getters for my tests representation of business entities and throws, when trying to access something that hadnt been saved earlier:

```ts
import { setWorldConstructor, World } from '@wdio/cucumber-framework';
import type { BatchAnalysis, Goal, Order, Sample, SampleType } from '../../../common/entityTypes';

interface WorldProps {
    orders?: Order[],
    samples?: Sample[],
    goals?: Goal[],
    batchAnalyses?: BatchAnalysis[],
    template?: string,
    sampleType?: SampleType
    addedParameterLabel?: string,
    methodLabel?: string,
    patientCode?: string
}

const tryGetAll = <T>(entries: T[] | undefined, displayName: string): T[] => 
{
    if (!entries || entries.length < 1) 
        throw new Error(`No ${displayName} found in World - Check earlier steps`);
    
    return entries;
};

const tryGetNth = <T>(entries: T[] | undefined, nthEntry: number, displayName: string): T => 
{
    const existingEntries = tryGetAll(entries, displayName);

    if (existingEntries.length < nthEntry) 
        throw new Error(`Can not get ${nthEntry}. ${displayName}! Only ${existingEntries.length} ${displayName} found in World - Check earlier steps`);
    
    return existingEntries[nthEntry - 1];
};

//all 'nthSomething' parameters refer to ordinals (1-based-indices)
export class SamplesWorld extends World 
{
    private props: WorldProps = {};

    constructor(options) { super(options); }
   
    getOrder(nthOrder: number) { return tryGetNth(this.props.orders, nthOrder, 'Order'); }

    set orders(orders: Order[]) { this.props.orders = orders; }

    get orders() { return tryGetAll(this.props.orders, 'Orders'); }

    addOrder(order: Order) 
    { 
        this.props.orders
            ? this.props.orders.push(order)
            : this.props.orders = [order];
    }

    // .. more pretty similar stuff
}

setWorldConstructor(SamplesWorld);
```

Now i pushed those things a level deeper and added my fixture/views to its top-level to inject playwrights bultin features. This seemed to be the most straightforward way for me:

```ts
export class SamplesWorld extends BddWorld 
{
    constructor(options: BddWorldOptions) 
    {
      super(options);
    }

    get testData() { return new SamplesWorldDataStorage(); }

    public open = () => this.page.goto('/');

    //app
    public get loginView() { return new LoginView(this.builtinFixtures, 'login-view'); }
    
    public get navigation() { return new Navigation(this.builtinFixtures, '#navigation'); }
    
    public get samplesView() { return new SamplesListView(this.builtinFixtures, 'samples-list-view'); }
    public get samplesCreateView() { return new SamplesCreateView(this.builtinFixtures, 'sample-create-view'); }
    public get samplesEditView() { return new SamplesEditView(this.builtinFixtures, 'sample-edit-view'); }

    // .. more similiar stuff
}

setWorldConstructor(SamplesWorld);
```

**Comment by vitalets** [comment]
I've added a section to docs [how you can use hooks together with `playwright-bdd`](https://vitalets.github.io/playwright-bdd/#/./hooks).
Please let me know is it helpful for you.

**Comment by gzuppl** [comment]
By now i have decided that i am not going to stay at "cucumber-style-steps" for much longer - I really need those hooks back and pw custom fixtures offer some other benefits i would like to make use of. 

The added section gives me a good starting point to then create individual hooks for the few specific tests, that require "special state reset". Thanks!

**Comment by vitalets** [comment]
> The added section gives me a good idea how to then create individual hooks for the few specific tests, that require "special state reset". Thanks!

It was my intention :) Let me know if there are any problems on that.

**Comment by vitalets** [comment]
Hi all! I've released hooks support in 5.5.0.
Here is a new [documentation for hooks](https://vitalets.github.io/playwright-bdd/#/./hooks).

I'll appreciate if try it and share your feedback 😊

**Comment by bj00rn** [comment]
I don't quite understand the documentation on hooks, are cucumber style hooks supported? Do I have to switch to using fixtures?

login.feature
```
Feature: Logging in to the application

    As a user I want to be able to log in to the application.

    Background:
        Given I'm on the log in page

    Scenario: Logging in with email
        When I log in with correct email and password
        Then I'm on the home page
````
steps.js
```
import {Given, When, Then, Before, After} from '@cucumber/cucumber'
...
Before(function () {
   // should run before every scenario
    console.log('before')
})

After(function () {
    // should run after every scenario
    console.log('after')
})
...
```

**Comment by bj00rn** [comment]
thanks for the quick reply! Does the Before({tags: '@unresponsive-api'}.. syntax work
i.e

```
   @unresponsive-api
    Scenario: Show error message when server is unresponsive
        Then I'm on the log in page with the error message "Server unavailable."

```

```
Before({tags: '@unresponsive-api'}, function () {
   // should run before every scenario tagged '@unresponsive-api'
    console.log('before')
});
```

Im struggling with and old codebase with a lot of tests. I would be super happy just get them to work rather than having to switch from tags to features.

**Comment by vitalets** [comment]
Yes, tags, timeout and name options are supported for Before / After hooks.

**Comment by swayf** [comment]
> Imagine a scenario with a step that requires authorization:
> 
> ```
> Feature: Some feature
> 
>     Scenario: Some scenario
>         Given I am an authorized user
> ```
> 
> So we need to sign-in user before running step and sign-out after the step.
> 
> **In Cucumber** I can add some tag (e.g. `@auth`) to that scenario and add Before and After conditional hooks:
> 
> ```
> Feature: Some feature
> 
>     @auth
>     Scenario: Some scenario
>         Given I am an authorized user
> ```
> 
> Hooks:
> 
> ```ts
> Before({tags: '@auth'}, async function () {
>   console.log('sign-in');
> });
> 
> After({tags: '@auth'}, async function () {
>   console.log('sign-out');
> });
> ```
> 
> **In Playwright** I can solve it with the following `auth` fixture:
> 
> ```ts
> export const test = base.extend({
>   auth: async ({}, use) => {
>     console.log('sign-in');
>     await use({ username: 'some user' });
>     console.log('sign-out');
>   }
> });
> ```
> 
> And just reference `auth` fixture in the step definition:
> 
> ```ts
> Given('I am an authorized user', async ({ auth }) => {
>   console.log('step for authorized user', auth.username);
> });
> ```

Is it possible to do it on "feature level" with hooks or some another way? 
Like we do sign-in once for the all scenarios of the feature.
Just to save some time, if it is enough to do some preparation only once per feature but not for every scenarios ..

**Comment by vitalets** [comment]
> Is it possible to do this on "feature level" with hooks or some another way?
Like we do sign-in once for the all scenarios of the feature.
Just to save some time, if it is enough to do some preparation only once per feature but not for every scenarios ..

@swayf 
I think worker-scoped fixture is the best solution. Feel free to open a separate issue with your case.
